### PR TITLE
V2.0 speedup rpn

### PR DESCRIPTION
**Pull request by yhcao6**
Speed up RPN head.

| Model             | Original Inf time (s/iter) | New Inf time (s/iter) |
|-------------------|----------------------------|-----------------------|
| Faster R-CNN R-50 | 16.0                       | 18.0                  |
| Mask R-CNN R-50   | 10.9                       | 11.9                  |
| Retinanet R-50    | 16                         |   16.1                    |

### CODE REVIEW

File: mmdet/models/anchor_heads/rpn_head.py [hunk 1]
```diff
@@ -79,30 +82,38 @@ def get_bboxes_single(self,
             rpn_bbox_pred = rpn_bbox_pred.permute(1, 2, 0).reshape(-1, 4)
             anchors = mlvl_anchors[idx]
             if cfg.nms_pre > 0 and scores.shape[0] > cfg.nms_pre:
-                _, topk_inds = scores.topk(cfg.nms_pre)
+                # _, topk_inds = scores.topk(cfg.nms_pre)
```
Comment (ZwwWayne) [inline mmdet/models/anchor_heads/rpn_head.py:85]: You may add comments here to indicate that sort is faster than topk.

File: mmdet/models/anchor_heads/rpn_head.py [hunk 2]
```diff
@@ -79,30 +82,38 @@ def get_bboxes_single(self,
...
-                                   self.target_stds, img_shape)
-            if cfg.min_bbox_size > 0:
-                w = proposals[:, 2] - proposals[:, 0]
-                h = proposals[:, 3] - proposals[:, 1]
-                valid_inds = torch.nonzero((w >= cfg.min_bbox_size) &
-                                           (h >= cfg.min_bbox_size)).squeeze()
+            mlvl_scores.append(scores)
+            mlvl_bbox_preds.append(rpn_bbox_pred)
+            mlvl_valid_anchors.append(anchors)
+            level_ids.append(
```
Comment (ZwwWayne) [inline mmdet/models/anchor_heads/rpn_head.py:94]: The function of `level_ids` here may be a little bit confusing at first glance until the users read line 118.
You may add comments here to demonstrate that boxes from different levels are sent to NMS separately, the level_ids act as labels of different classes to use the `batched_nms` for acceleration.

File: mmdet/models/anchor_heads/rpn_head.py [hunk 3]
```diff
@@ -79,30 +84,39 @@ def get_bboxes_single(self,
...
+                torch.full((scores.size(0), ),
```
Comment (hellock) [inline mmdet/models/anchor_heads/rpn_head.py:98]: Use `scores.new_full()` instead.

File: mmdet/models/anchor_heads/rpn_head.py [hunk 4]
```diff
@@ -79,30 +84,39 @@ def get_bboxes_single(self,
...
-            proposals = proposals[:cfg.max_num, :]
-        else:
-            scores = proposals[:, 4]
-            num = min(cfg.max_num, proposals.shape[0])
-            _, topk_inds = scores.topk(num)
-            proposals = proposals[topk_inds, :]
-        return proposals
+                ids = ids[valid_inds]
+
+        nms_cfg = dict(type='nms', iou_thr=cfg.nms_thr)
```
Comment (hellock) [inline mmdet/models/anchor_heads/rpn_head.py:120]: Add a `TODO` here to remove the hard coded nms type.

File: mmdet/ops/nms/nms_wrapper.py [hunk 5]
```diff
@@ -116,3 +116,23 @@ def soft_nms(dets, iou_thr, method='linear', sigma=0.5, min_score=1e-3):
     else:
         return new_dets.numpy().astype(dets.dtype), inds.numpy().astype(
             np.int64)
+
+
+def batched_nms(bboxes, scores, idxs, nms_cfg):
```
Comment (hellock) [inline mmdet/ops/nms/nms_wrapper.py:121]: Add a standard docstring.

File: mmdet/ops/nms/nms_wrapper.py [hunk 6]
```diff
@@ -116,3 +116,36 @@ def soft_nms(dets, iou_thr, method='linear', sigma=0.5, min_score=1e-3):
...
+    Modified from https://github.com/pytorch/vision/blob
+    /505cd6957711af790211896d32b40291bea1bc21/torchvision/ops/boxes.py#L39.
+    In order to perform NMS independently per class, we add an offset to all
+    the boxes. The offset is dependent only on the class idx, and is large
+    enough so that boxes from different classes do not overlap.
+
+    Arguments:
+        bboxes (torch.Tensor): bboxes in shape (N, 4).
+        scores (torch.Tensor): scores in shape (N, ).
+        idxes (torch.Tensor): each index value correspond to a category, and
```
Comment (hellock) [inline mmdet/ops/nms/nms_wrapper.py:133]: - Rename it to `inds`, which is short for `indices` (plural form of index).
- `inds` does not necessarily correspond to categories.

**Pull request merged**